crates/services/integration/src/komga.rs

```diff
@@ -0,0 +1,478 @@
+        /// in the url
+        ///
+        /// # Arguments
+        ///
+        /// * `url`: The url to extact from
+        ///
+        /// returns: The ID number if the extraction is successful
+        fn extract_id(&self, url: String) -> Option<String> {
+            if let Ok(parsed_url) = Url::parse(&url) {
+                parsed_url
+                    .path_segments()
+                    .and_then(|segments| segments.collect::<Vec<_>>().get(1).cloned())
+                    .map(String::from)
+            } else {
+                None
+            }
+        }
+
+        /// Extracts the list of providers with a MediaSource,ID Tuple
+        ///
+        /// Currently only works for myanimelist and anilist as mangaupdates doesn't store
+        /// the ID in the url
+        ///
+        /// Requires that the metadata is stored with the label anilist or myanimelist
+        /// other spellings wont work
+        ///
+        /// returns: list of providers with a MediaSource, ID Tuple
+        pub fn find_providers(&self) -> Vec<(MediaSource, Option<String>)> {
+            let mut provider_links = vec![];
+            for link in self.links.iter() {
+                // NOTE: manga_updates doesn't work here because the ID isn't in the url
+                let source = match link.label.to_lowercase().as_str() {
+                    "anilist" => MediaSource::Anilist,
+                    "myanimelist" => MediaSource::Mal,
+                    _ => continue,
+                };
+
+                let id = self.extract_id(link.url.clone());
+                provider_links.push((source, id));
+            }
+
+            provider_links.sort_by_key(|a| a.1.clone());
+            provider_links
+        }
+    }
+
+    #[derive(Debug, Serialize, Deserialize)]
+    #[serde(rename_all = "camelCase")]
+    pub struct Item {
+        pub id: String,
+        pub name: String,
+        pub books_count: Decimal,
+        pub books_read_count: Option<i32>,
+        pub books_unread_count: Decimal,
+        pub metadata: Metadata,
+    }
+
+    #[derive(Debug, Serialize, Deserialize)]
+    #[serde(rename_all = "camelCase")]
+    pub struct Response {
+        pub content: Vec<Item>,
+    }
+}
+
+mod komga_events {
+    use super::*;
+
+    #[derive(Debug, Serialize, Deserialize, Clone)]
+    #[serde(rename_all = "camelCase")]
+    pub struct Data {
+        pub book_id: String,
+        pub user_id: String,
+    }
+}
+
+struct KomgaEventHandler {
+    task: OnceLock<()>,
+    receiver: Mutex<Option<UnboundedReceiver<komga_events::Data>>>,
+}
+
+impl KomgaEventHandler {
+    pub const fn new() -> Self {
+        Self {
+            task: OnceLock::new(),
+            receiver: Mutex::new(None),
+        }
+    }
+
+    pub fn get_receiver(&self) -> &Mutex<Option<UnboundedReceiver<komga_events::Data>>> {
+        &self.receiver
+    }
+
+    pub fn get_task(&self) -> &OnceLock<()> {
+        &self.task
+    }
+}
+
+impl IntegrationService {
```

Comment:
I think a better way to structure this would be to create a `KomgaService` which exposes a `progress` function. Then `IntegrationService` will construct this service and hand off the actual sync to this function.

The same for all the other integrations.

The way you have implemented it is fine, but it does pollute `IntegrationService` with additional functions like `process_events` etc. which I don't like.

Reply:
I agree Ill see what I can do

Reply:
Updated to use a Integration trait that provides a progress function so all Yank Integrations can implement a progress function. We maybe should rename it but im not sure what to call the trait.

Reply:
Name is fine. Could you migrate the other integrations as well?